[inductor] If a kernel contains bucketize, try using config with num_elements_per_warp=32

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104456

In binary search triton implementations, (#104007) num_elements_per_warp=32 performs a lot better than larger values.

This PR adds an autotuning config option for this purpose. But since autotuning can affect compile times and this config isn't generally useful, we only try this config if bucketize is present. This is done by adding an extra field to triton_meta which is used by the pointwise autotuning 

Performance: reused https://gist.github.com/davidberard98/066fd2115f59f5889ef61e4527d1eba5. 

Before:
```
Eager 0.30088499188423157 ms
PT2   0.9296960234642029 ms
```

After:
```
Eager 0.3011910021305084 ms
PT2   0.22977299988269806 ms
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

Differential Revision: [D47237103](https://our.internmc.facebook.com/intern/diff/D47237103)

davidberard98 committed:
[inductor] Add pointwise autotuning config where num_elements_per_warp is 32

In binary search triton implementations, () num_elements_per_warp=32 performs a lot better than larger values.

This PR adds an autotuning config option for this purpose.

[ghstack-poisoned]

eellison commented:
Could we try to add some heuristics so that we don't benchmark all three configs every time?  Each config adds 7s of compilation time to TIMM models.

davidberard98 commented:
Oh, didn't realize it had such a big impact. I'll work on doing something like that.

davidberard98 committed:
Update on "[inductor] Add pointwise autotuning config where num_elements_per_warp=32"


In binary search triton implementations, (#104007) num_elements_per_warp=32 performs a lot better than larger values.

This PR adds an autotuning config option for this purpose.

I benchmarked #104007 with and without this change on a 16MB pointwise kernel. This change reduces the latency from 1ms to 0.35ms.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

aakhundov commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -706,6 +733,10 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
     bs = max(256, min(numel // 128, 1024))
 
+    hinted_configs = autotune_hints_to_configs(
+        meta.get("autotune_hints", {}), size_hints, bs
```
Super-nit: `{}` is a dict, whereas `autotune_hints_to_configs` expects a `Set[AutotuneHint]` as the first argument. The iteration behavior is the same, but maybe let's change `{}` to `set()` for clarity (personally, I was confused with it being an empty dict here, as I didn't think keys are required).

aakhundov reviewed (commented):
Thanks @davidberard98! Looks great, left a few (mostly exploratory) comments.

davidberard98 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -723,6 +754,7 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
                 [
                     triton_config(size_hints, bs, num_elements_per_warp=256),
                     triton_config(size_hints, bs // 2, num_elements_per_warp=64),
+                    *hinted_configs,
```
afaik x*y*z == number of elements, and then we divide by num_elements_per_warp to get the number of warps.

when x -> x/2 and num_elements_per_warp -> num_elements_per_warp/2, then number of warps doesn't change.

In this case I put `bs // 4` because there appears to be a max number of warps of 8 in the triton_config implementation; so we need x*y*z <= 256 in order to actually get num_elements_per_warp=32.

eellison replied:
It's a bit confusing. add comment ?

davidberard98 replied:
Added a comment in `def triton_config(...`

davidberard98 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -53,6 +53,33 @@ class HeuristicType(Enum):
+            configs.append(
+                triton_config(
+                    size_hints,
+                    *[1 for _ in range(len(size_hints) - 1)],
+                    bs // 4,
```
Added all the configs (i.e. [bs, 1, 1], [1, bs, 1], [1, 1, bs] for rank-3 `size_hints`)

davidberard98 committed:
Update on "[inductor] If a kernel contains bucketize, try using config with num_elements_per_warp=32"


In binary search triton implementations, (#104007) num_elements_per_warp=32 performs a lot better than larger values.

This PR adds an autotuning config option for this purpose. But since autotuning can affect compile times and this config isn't generally useful, we only try this config if bucketize is present. This is done by adding an extra field to triton_meta which is used by the pointwise autotuning 

Performance: reused https://gist.github.com/davidberard98/066fd2115f59f5889ef61e4527d1eba5. 

Before:
```
Eager 0.30088499188423157 ms
PT2   0.9296960234642029 ms
```

After:
```
Eager 0.3011910021305084 ms
PT2   0.22977299988269806 ms
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47237103).

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -53,6 +53,45 @@ class HeuristicType(Enum):
+def autotune_hints_to_configs(
+    hints: Set[AutotuneHint], size_hints, block_size
+) -> List[Config]:
+    configs = []
```
docstring for this function ?

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -53,6 +53,45 @@ class HeuristicType(Enum):
+
+    for hint in hints:
+        if hint == AutotuneHint.ELEMENTS_PER_WARP_32:
+            if len(size_hints) == 1:
```
Would be more concise as :

```xyz_options = [tuple([ block_size // 4 if j == i else 1 for j in range(len(size_hints))]) for i in range(len(size_hints))]```

yours is probably more readable, though

davidberard98 replied:
had something like this before but I decided to expand it for readability and because the configs defined below also follow this sort of verbose style

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -53,6 +53,45 @@ class HeuristicType(Enum):
+                    (1, 1, block_size // 4),
+                )
+            for xyz in xyz_options:
+                configs.append(
```
This is adding one config per dimension where previously you just added one. Is that intentional ? needed ?

davidberard98 replied:
Intentional, yes; needed, idk.

davidberard98 replied:
did some experiments, it seems like it _is_ useful to have multiple configs (e.g. (1, 256) and (256, 1)) - the best option will sometimes depend on the shape of size_hints

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -53,6 +53,45 @@ class HeuristicType(Enum):
+        if hint == AutotuneHint.ELEMENTS_PER_WARP_32:
+            if len(size_hints) == 1:
+                xyz_options = [(block_size // 4,)]
+            elif len(size_hints) == 2:
```
dont think `len(size_hints) != 1` is currently exercised in tests

davidberard98 replied:
any suggestions how to trigger this? I haven't been able to figure out a way to make this run

eellison replied:
Try a channels last tensor ?

davidberard98 replied:
Thanks - added a test in test_torchinductor.py for now, lmk if that is sufficient; if not I can try to add a separate test file for pointwise autotuning

eellison reviewed (commented):
nice ! couple comments

davidberard98 committed:
Update on "[inductor] If a kernel contains bucketize, try using config with num_elements_per_warp=32"


In binary search triton implementations, (#104007) num_elements_per_warp=32 performs a lot better than larger values.

This PR adds an autotuning config option for this purpose. But since autotuning can affect compile times and this config isn't generally useful, we only try this config if bucketize is present. This is done by adding an extra field to triton_meta which is used by the pointwise autotuning 

Performance: reused https://gist.github.com/davidberard98/066fd2115f59f5889ef61e4527d1eba5. 

Before:
```
Eager 0.30088499188423157 ms
PT2   0.9296960234642029 ms
```

After:
```
Eager 0.3011910021305084 ms
PT2   0.22977299988269806 ms
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

Differential Revision: [D47237103](https://our.internmc.facebook.com/intern/diff/D47237103)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47237103).

eellison reviewed (approved):
LGTM - if compilation time is high this is maybe something to revisit, to reduce number of configs / improve heuristics.
